### PR TITLE
THW: Add binary encoding/decoding functions

### PR DESCRIPTION
**Pull request by neilalexander**
Signed-off-by: Neil Twigg <neil@nats.io>

### CODE REVIEW

**Review by derekcollison — CHANGES_REQUESTED** [review]

File: server/thw/thw.go [hunk 1]
```diff
@@ -194,3 +197,49 @@ func (hw *HashWheel) GetNextExpiration(before int64) int64 {
...
+// AppendEncode writes out the contents of the THW into a binary format
+// and appends it onto the end of the provided slice, returning the new
+// slice.
+func (hw *HashWheel) AppendEncode(b []byte) []byte {
+	b = append(b, 1)                        // Magic version
+	b = binary.BigEndian.AppendUint64(b, 0) // Entry count
+	var count uint64
+	for _, slot := range hw.wheel {
+		for v, ts := range slot.entries {
+			b = binary.AppendVarint(b, ts)
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply server/thw/thw.go:210]: The only time it's slower than `Put()` is where a reallocation is involved, but we don't know the needed buffer size up front and we don't currently have the functions available in the stdlib to work out how big uvarints/varints will be before we encode them, so we pretty much need to rely on reallocation behaviour here to grow the slice as needed.

File: server/thw/thw.go [hunk 2]
```diff
@@ -194,3 +197,49 @@ func (hw *HashWheel) GetNextExpiration(before int64) int64 {
 	}
 	return math.MaxInt64
 }
+
+// AppendEncode writes out the contents of the THW into a binary format
+// and appends it onto the end of the provided slice, returning the new
+// slice.
+func (hw *HashWheel) AppendEncode(b []byte) []byte {
+	b = append(b, 1)                        // Magic version
+	b = binary.BigEndian.AppendUint64(b, 0) // Entry count
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply server/thw/thw.go]: I had thought originally of just doing this outside, but have added a stamp parameter instead.

[the remark this answers is not in the document]
Comment (neilalexander) [reply server/thw/thw.go]: Done.

File: server/thw/thw.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/thw/thw.go:212]: More reason to track numEntries in thw I think..

File: server/thw/thw.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/thw/thw.go]: We should add that in.

[the remark this answers is not in the document]
Comment (derekcollison) [reply server/thw/thw.go]: The THW adds in sequences, so maybe highSeq or lastSeq.

[the remark this answers is not in the document]
Comment (neilalexander) [reply server/thw/thw.go]: Done!

File: server/thw/thw.go [hunk 3]
```diff
@@ -194,3 +203,49 @@ func (hw *HashWheel) GetNextExpiration(before int64) int64 {
 	}
 	return math.MaxInt64
 }
+
+// AppendEncode writes out the contents of the THW into a binary snapshot
+// and returns it. The high seq number is included in the snapshot and will
+// be returned on decode.
+func (hw *HashWheel) Encode(highSeq uint64) []byte {
+	b := make([]byte, 0, 17+(hw.count*(2*binary.MaxVarintLen64)))
```
Comment (derekcollison) [inline server/thw/thw.go:211]: Make 17 a const?

File: server/thw/thw.go [hunk 4]
```diff
@@ -194,3 +203,49 @@ func (hw *HashWheel) GetNextExpiration(before int64) int64 {
...
+}
+
+// Decode snapshots a binary-encoded THW and replaces the contents of this
+// THW with them. Returns the high seq number from the snapshot.
+func (hw *HashWheel) Decode(b []byte) (uint64, error) {
+	if len(b) < 17 {
+		return 0, io.ErrShortBuffer
+	}
+	if b[0] != 1 {
+		return 0, fmt.Errorf("unknown magic %d", b[0])
```
Comment (derekcollison) [inline server/thw/thw.go:231]: Add to know errors?

File: server/thw/thw.go (hunk 3 above)
Comment (neilalexander) [reply server/thw/thw.go]: Done.

File: server/thw/thw.go (hunk 4 above)
Comment (neilalexander) [reply server/thw/thw.go]: Done.

Comment (derekcollison) [reply server/thw/thw.go]: nvm I see its variable. But we could make it a fixed error.

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**